src/Adapter/Product/CommandHandler/SetProductCustomizationFieldsHandler.php

```diff
@@ -0,0 +1,190 @@
+     * @param CustomizationFieldProvider $customizationFieldProvider
+     * @param ProductUpdater $productUpdater
+     */
+    public function __construct(
+        AddCustomizationFieldHandlerInterface $addCustomizationFieldHandler,
+        CustomizationFieldPersister $customizationFieldPersister,
+        CustomizationFieldDeleterInterface $customizationFieldDeleter,
+        CustomizationFieldProvider $customizationFieldProvider,
+        ProductUpdater $productUpdater
+    ) {
+        $this->addCustomizationFieldHandler = $addCustomizationFieldHandler;
+        $this->customizationFieldDeleter = $customizationFieldDeleter;
+        $this->customizationFieldPersister = $customizationFieldPersister;
+        $this->customizationFieldProvider = $customizationFieldProvider;
+        $this->productUpdater = $productUpdater;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * Creates, updates or deletes customization fields depending on differences of existing and provided fields
+     */
+    public function handle(SetProductCustomizationFieldsCommand $command): array
+    {
+        $deletableFieldIds = $this->getDeletableFieldIds($command);
+
+        foreach ($command->getCustomizationFields() as $customizationFieldDTO) {
+            if ($customizationFieldDTO->getCustomizationFieldId()) {
+                $this->handleUpdate($customizationFieldDTO);
+            } else {
+                $this->handleCreation($command->getProductId(), $customizationFieldDTO);
+            }
+        }
+
+        $this->deleteCustomizationFields($deletableFieldIds);
+        $product = $this->getProduct($command->getProductId());
+        $this->productUpdater->refreshProductCustomizabilityProperties($product);
+
+        return array_map(function (int $customizationFieldId): CustomizationFieldId {
+            return new CustomizationFieldId($customizationFieldId);
+        }, $product->getNonDeletedCustomizationFieldIds());
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param CustomizationFieldDTO $customizationFieldDTO
+     */
+    public function handleCreation(ProductId $productId, CustomizationFieldDTO $customizationFieldDTO): void
+    {
+        $customizationField = new CustomizationField();
+        $customizationField->id_product = $productId->getValue();
+        $customizationField->type = $customizationFieldDTO->getType();
+        $customizationField->required = $customizationFieldDTO->isRequired();
+        $customizationField->name = $customizationFieldDTO->getLocalizedNames();
+        $customizationField->is_module = $customizationFieldDTO->isAddedByModule();
+
+        $this->customizationFieldPersister->add($customizationField);
+    }
+
+    /**
+     * @param CustomizationFieldDTO $customizationFieldDTO
+     *
+     * @throws CoreException
+     */
+    private function handleUpdate(CustomizationFieldDTO $customizationFieldDTO): void
+    {
+        $fieldId = $customizationFieldDTO->getCustomizationFieldId();
```

Comment:
It's too bad we need to instanciate the VO from the DTO just to pass it as an argument to the persister?

Maybe it's a dumb question but can't DTO use VOs? like `$customizationFieldDTO->getCustomizationFieldId()` could return a `CustomizationFieldId` directly?

Reply:
I don't think so.
It suppose to just transfer data without doing anything else.
In other words it prevents us from using array, nothing more.

Reply:
Mmm, then maybe DTO could be used in place of the array in our Persister/Updater services 🤔 Assuming they would have a common interface, like allowing to detect if such field has been set or not for partial update (maybe based on Symfony ParameterBag or something)
This is just a thought of improvement not to be handled right now of course ^^

Reply:
It would mean separate DTO for each ObjectModel and thats already confusing enough :smile: